core/src/main/java/com/uber/m3/tally/sanitizers/ValidCharacters.java

```diff
@@ -0,0 +1,110 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.uber.m3.tally.sanitizers;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * ValidCharacters is a collection of valid characters.
+ */
+public class ValidCharacters {
+
+    /**
+     * DEFAULT_REPLACEMENT_CHARACTER is the default character used for replacements.
+     */
+    public static char DEFAULT_REPLACEMENT_CHARACTER = '_';
+
+    /**
+     * ALPHANUMERIC_RANGE is the range of alphanumeric characters.
+     */
+    public static final List<CharRange> ALPHANUMERIC_RANGE =
+        Arrays.asList(
+            CharRange.of('a', 'z'),
+            CharRange.of('A', 'Z'),
+            CharRange.of('0', '9'));
+
+    /**
+     * UNDERSCORE_CHARACTERS contains the underscore character.
+     */
+    public static final List<Character> UNDERSCORE_CHARACTERS = Collections.singletonList('_');
+
+    /**
+     * UNDERSCORE_DASH_CHARACTERS contains the underscore and dash characters.
+     */
+    public static final List<Character> UNDERSCORE_DASH_CHARACTERS = Arrays.asList('_', '-');
+
+    /**
+     * UNDERSCORE_DASH_DOT_CHARACTERS contains the underscore, dash and dot characters.
+     */
+    public static final List<Character> UNDERSCORE_DASH_DOT_CHARACTERS = Arrays.asList('_', '-', '.');
+
+    private final List<CharRange> ranges;
+    private final List<Character> characters;
+
+    private ValidCharacters(List<CharRange> ranges, List<Character> characters) {
+        this.ranges = ranges != null ? ranges : Collections.emptyList();
+        this.characters = characters != null ? characters : Collections.emptyList();
+    }
+
+    public static ValidCharacters of(List<CharRange> ranges, List<Character> characters) {
+        return new ValidCharacters(ranges, characters);
+    }
+
+    Function<String, String> sanitizeStringFunc(char repChar) {
+        return value -> {
+            StringBuilder buffer = null;
+
+            for (int i = 0; i < value.length(); i++) {
+                char ch = value.charAt(i);
+
+                // first check if the provided character is valid
+                boolean validCurr =
+                    ranges.stream().anyMatch(range -> ch >= range.low() && ch <= range.high())
+                        || characters.stream().anyMatch(character -> character.equals(ch));
```

Comment:
`.contains` seems more readable to me and more efficient if you use `Set`

Reply:
Good idea. will do it